### PR TITLE
Planet Four fan tool

### LINKED ISSUE #4329: Fan Tool for Planet Four migration to panoptes
Separated out from previous ticket to focus on the main feature needed to transition P4 from Ouroboros to Panoptes:

- [ ] a fan tool: the ice cream cone drawing tool where you can drag from a starting point outward. We did Panoptes triangle tool isn't close to this.  
fans-3

### PR DESCRIPTION
**Pull request by eatyourgreens**
Add the 'ice cream cone' drawing tool from Planet Four. Tool annotation params will be radius, rotation angle and spread angle.

ToDo:
- [x] Create a draggable radial line tool.
- [x] Add a draggable perpendicular line to represent spread.
- [x] Render the final fan shape as a triangle and semi-circle.
- [x] Add the `deleteIfOutOfBounds` check.
- [x] Add the tool to the Project Builder.
- [x] Add tests.

Staging branch URL: https://fan-tool.pfe-preview.zooniverse.org/dev/classifier
(Choose 'Draw stuff', 'Next' then 'Fan tool' at the bottom of the list.

Fixes #4329.

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
@mschwamb The fan tool is available on staging now.
https://fan-tool.pfe-preview.zooniverse.org/dev/classifier
(Choose 'Draw stuff', 'Next' then 'Fan tool' at the bottom of the list.)

Screenshot of the fan tool in the PFE dev classifier.

**Comment by eatyourgreens** [comment]
The code needs tests (particularly the methods that use trig functions, I think) but this should be good to go, otherwise.

**Comment by eatyourgreens** [comment]
There's a weird bug around rotation angles of 270º or so, where the spread angle suddenly jumps from around 60º to around 650º, as you drag the upper drag handle.
Screenshot, showing an inverted cone shape with a spread angle larger than 360º.

**Comment by CKrawczyk** [comment]
That but seems to happen when the upper handle passes from `+y` to `-y`.  A similar thing happens when you rotate it so the to around -140 and drag the bottom handle down.

**Comment by eatyourgreens** [comment]
I'm going to try writing some tests for those cases.

**Comment by eatyourgreens** [comment]
@mschwamb any thoughts on reasonable limits for the angles that this tool measures?

**Comment by mschwamb** [comment]
180 degrees /semicircle should be the largest extent of the fan angle. The shape looks weird in panoptes when you go beyond 180 degrees

screen shot 2018-09-11 at 3 26 27 am

screen shot 2018-09-11 at 3 24 42 am

**Comment by CKrawczyk** [comment]
The bug from above happens when `cursorAngle` in the `handleSpread` function passes from 180 to -180.

File: app/classifier/drawing-tools/components/Fan/fan.jsx [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    const rotation = Fan.getCursorAngle({ x, y }, mark);
+    const newMark = Object.assign({}, mark, { radius, rotation });
+    this.props.onChange(newMark);
+  }
+
+  handleSpread(e) {
+    const { mark, getEventOffset } = this.props;
+    const { x, y } = getEventOffset(e);
+    const cursorAngle = Fan.getCursorAngle({ x, y }, mark);
+    const spread = 2 * Math.abs(cursorAngle - mark.rotation);
```
Comment (CKrawczyk) [inline app/classifier/drawing-tools/components/Fan/fan.jsx]: The angle difference here needs to take into account when `cursorAngle` passes from `180` to `-180`.  This can be done with a function similar to:

```js
let spread = Math.abs(cursorAngle - mark.rotation);
if (spread > 180) {
    spread = 360 - spread;
}
spread *= 2;
```

Comment (eatyourgreens) [reply app/classifier/drawing-tools/components/Fan/fan.jsx]: Good point. I've started writing tests, so I'll add a test for this, then fix the code to pass the test.

Comment (eatyourgreens) [reply app/classifier/drawing-tools/components/Fan/fan.jsx]: I've added some tests but they don't trip this bug. 🤔 

I've also added a failing test for a maximum spread angle of 180º.

**Comment by eatyourgreens** [comment]
Should all be fixed now, but I'd be a lot happier if there was a test that caught that bug with `cursorAngle`.

File: app/classifier/drawing-tools/components/Fan/fan.spec.jsx [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            mark.rotation = mark.rotation < -180 ? 360 + mark.rotation : mark.rotation;
+            wrapper.setProps({ mark });
+            wrapper.instance().handleSpread(cursor);
+            expect(onChange.callCount).to.equal(1);
+            const newMark = onChange.getCall(0).args[0];
+            expect(newMark.spread).to.equal(180);
+            onChange.resetHistory();
+          });
+        });
+      })
```
Comment (CKrawczyk) [inline app/classifier/drawing-tools/components/Fan/fan.spec.jsx]: Missing `;`

File: app/classifier/drawing-tools/components/Fan/fan.spec.jsx [hunk 3]
```diff
@@ -0,0 +1,173 @@
+import React from 'react';
+import sinon from 'sinon';
+import { shallow } from 'enzyme';
+import { expect } from 'chai';
+import DrawingToolRoot from '../../root';
```
Comment (CKrawczyk) [inline app/classifier/drawing-tools/components/Fan/fan.spec.jsx]: Is this import used?

Comment (eatyourgreens) [reply app/classifier/drawing-tools/components/Fan/fan.spec.jsx]: Or maybe not.

**Comment by CKrawczyk** [comment]
The only thing I would flag up about the 0 to 360 change is it is different than the angle formats for our other drawing tools.  As this tool is likely to be used for a small number of projects I don't think it is that big a deal.

From an aggregation point of view it would be nice if all the drawing tools followed a single standard instead of each one using a different one.

Just to provide context for this here is how the angle is defined across all of our other tools:
+ ellipse: -180 to 180, positive angle rotates CCW, 0 off of +x axis
+ triangle: -180 to 180, positive angle rotates CCW, 0 off of -y axis
+ rotating rectangle: -180 to 180, positive angle rotates CW, 0 off of +x axis
+ Fan: 0 to 360, positive angle rotates CW, 0 off of +x axis

**Comment by eatyourgreens** [comment]
That's a good point about documenting the drawing tools. #4478 might help, or even just a readme file in the `drawing-tools` directory.

#4329 doesn't specify how the fan tool angles should be measured, so I made my best guess here.

**Comment by michaelaye** [comment]
I'm not seeing the distance (=total length) value being mentioned that we currently measure? It should be stored as well.

**Comment by eatyourgreens** [comment]
Is that different from the mark radius?

**Comment by michaelaye** [comment]
I don't know, the fan tool never had a radius, so I'm not sure what "radius" would mean. Is there a graphic somewhere?

**Comment by eatyourgreens** [comment]
The staging link, above, let’s you play with the fan tool and see the annotations it produces. Now that this PR is merged to master, the fan’s also available on https://master.pfe-preview.zooniverse.org/dev/classifier so that you can try it out. 

There’s also tests in the code changes for this PR. If you’re comfortable reading code, they have cases for setting mark size, rotation and spread.

**Comment by eatyourgreens** [comment]
This comment has example annotations from a version earlier this week. https://github.com/zooniverse/Panoptes-Front-End/pull/4875#issuecomment-420210448

The current version measures rotation from 0 to 360° and spread is the full angle at the point of the triangle, rather than half the angle.

**Comment by eatyourgreens** [comment]
No problem #4329 didn’t go into detail about how the tool worked so I guessed how it should work here. If you’d like any changes, feel free to open a GitHub issue.
